numpy/core/src/multiarray/item_selection.c

```diff
@@ -1749,6 +1750,284 @@ PyArray_LexSort(PyObject *sort_keys, int axis)
+    char *indbuffer = PyDataMem_NEW(N * sizeof(npy_intp));
+    if (indbuffer == NULL) {
+       goto fail;
+    }
+    if (needcopy) {
+        char *valbuffer;
+        int *swaps;
+        npy_intp valbufsize = N * maxelsize;
+        if (NPY_UNLIKELY(valbufsize) == 0) {
+            valbufsize = 1;  /* Ensure allocation is not empty */
+        }
+
+        valbuffer = PyDataMem_NEW(valbufsize);
+        if (valbuffer == NULL) {
+            PyDataMem_FREE(indbuffer);
+            goto fail;
+        }
+        swaps = malloc(NPY_LIKELY(n > 0) ? n * sizeof(int) : 1);
+        if (swaps == NULL) {
+            PyDataMem_FREE(valbuffer);
+            PyDataMem_FREE(indbuffer);
+            goto fail;
+        }
+
+        for (j = 0; j < n; j++) {
+            swaps[j] = PyArray_ISBYTESWAPPED(mps[j]);
+        }
+        while (size--) {
+            iptr = (npy_intp *)indbuffer;
+            for (i = 0; i < N; i++) {
+                *iptr++ = i;
+            }
+            for (j = 0; j < n; j++) {
+                int rcode;
+                elsize = PyArray_DESCR(mps[j])->elsize;
+                astride = PyArray_STRIDES(mps[j])[axis];
+                argsort = PyArray_DESCR(mps[j])->f->argsort[NPY_STABLESORT];
+                if(argsort == NULL) {
+                    argsort = npy_atimsort;
+                }
+                _unaligned_strided_byte_copy(valbuffer, (npy_intp) elsize,
+                                             its[j]->dataptr, astride, N, elsize);
+                if (swaps[j]) {
+                    _strided_byte_swap(valbuffer, (npy_intp) elsize, N, elsize);
+                }
+                rcode = argsort(valbuffer, (npy_intp *)indbuffer, N, mps[j]);
+                if (rcode < 0 || (PyDataType_REFCHK(PyArray_DESCR(mps[j]))
+                            && PyErr_Occurred())) {
+                    PyDataMem_FREE(valbuffer);
+                    PyDataMem_FREE(indbuffer);
+                    free(swaps);
+                    goto fail;
+                }
+                PyArray_ITER_NEXT(its[j]);
+            }
+            char *slicebuffer = PyDataMem_NEW(N*aelsize);
+            _unaligned_strided_byte_copy(slicebuffer, aelsize, sit->dataptr,
+                                         selstride, N, aelsize);
+
+            iptr = (npy_intp *)indbuffer;
+            char *sortedbuffer = PyDataMem_NEW(N*aelsize);
+            if(sortedbuffer == NULL) {
+                PyDataMem_FREE(valbuffer);
+                PyDataMem_FREE(indbuffer);
+                goto fail;
+            }
+            for(i=0;i<N;i++){
+                int index = *iptr;
+                char *startOld = slicebuffer + (index*aelsize);
+                memmove(sortedbuffer+(i*aelsize), startOld, aelsize);
+                iptr++;
+            }
+            _unaligned_strided_byte_copy(sit->dataptr, selstride, sortedbuffer,
+                                        aelsize, N, aelsize);
```

Comment:
So this is not much faster, but I still like it a bit more, plus there is slightly more opportunity to speed it up if we ever want to, and less memory overhead.

Can't you do this in the for-loop directly?  Maybe we could also make a small helper for it, since I think it should look identical in both paths of the `needscopy` if/else.

Reply:
Made the change on this